### ISSUE TITLE
Improving and exposing the HTTP parts

### ISSUE THREAD

**Issue by vktr** [arvidn/libtorrent#5839]
A long while ago there was talk about moving to Boost.Beast for the HTTP client parts. Is this still relevant and would you accept a PR for this? That's the "improving" part.

Also, in PicoTorrent we do some HTTP requests to check for application updates etc. For us (as users of libtorrent) it would be nice to have access to the same HTTP client as libtorrent uses to make requests to share proxy settings etc. Does that make sense?

**Comment by arvidn** [comment]
Yes, I'm still interested in moving to boost.beast, and away from my home-grown `http_parser`. There's also an `http_connection` class that's used by `upnp.cpp` and by `http_tracker_connection.cpp`. With the tracker connection needing a lot of special logic, the generalized `http_connection` needs a lot of customization points (i.e. complexity) which I think is unwarranted.

I think a boost.beast conversion should aim to make `http_tracker_connection` just use boost.beast, and drop its dependency on the `http_connection` class (to eventually remove it).

Another improvement that should at least be kept in mind with a change like this is to be able to queue up additional requests to an `http_tracker_connection` object. If you have many torrents that share a tracker, you can save a *lot* of time, CPU and bandwidth by pipelining the requests on the same connection. I don't think this needs to be implemented in the same patch or PR, but it should be kept in mind.

**Comment by arvidn** [comment]
I'm a bit hesitant to expose the HTTP facilities. I understand the temptation, but it would restrict the implementation freedom in libtorrent, for the worse I think. It wouldn't be easy to transition to boost.beast, at least not while at the same time simplify the code.
